src/main/cpp/src/parse_uri.cu

```diff
@@ -23,279 +23,743 @@
 #include <cudf/scalar/scalar_factories.hpp>
 #include <cudf/strings/detail/strings_children.cuh>
 #include <cudf/strings/detail/utilities.cuh>
+#include <cudf/strings/string_view.cuh>
 
 #include <rmm/cuda_stream_view.hpp>
 #include <rmm/exec_policy.hpp>
 
 #include <memory>
+#include <tuple>
 
 namespace spark_rapids_jni {
 
 using namespace cudf;
 
 namespace detail {
+
+struct uri_parts {
+  string_view scheme;
+  string_view host;
+  string_view authority;
+  string_view path;
+  string_view fragment;
+  string_view query;
+  string_view userinfo;
+  string_view port;
+  string_view opaque;
+  bool valid{false};
+};
+
+enum URI_chunks { PROTOCOL, HOST, AUTHORITY, PATH, QUERY, USERINFO };
+
+enum chunk_validity { VALID, INVALID, FATAL };
+
 namespace {
 
-// utility to validate a character is valid in a URI
-constexpr bool is_valid_character(char ch, bool alphanum_only)
+// some parsing errors are fatal and some parsing errors simply mean this
+// thing doesn't exist or is invalid. For example, just because 280.0.1.16 is
+// not a valid IPv4 address simply means if asking for the host the host is null
+// but the authority is still 280.0.1.16 and the uri is not considered invalid.
+// By contrast, the URI https://[15:6:g:invalid] will not return https for the
+// scheme and is considered completely invalid.
+
+constexpr bool is_alpha(char c) { return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z'); }
+
+constexpr bool is_numeric(char c) { return c >= '0' && c <= '9'; }
+
+constexpr bool is_alphanum(char c) { return is_alpha(c) || is_numeric(c); }
+
+constexpr bool is_hex(char c)
 {
-  if (alphanum_only) {
-    if (ch >= '-' && ch <= '9' && ch != '/') return true;  // 0-9 and .-
-    if (ch >= 'A' && ch <= 'Z') return true;               // A-Z
-    if (ch >= 'a' && ch <= 'z') return true;               // a-z
-  } else {
-    if (ch >= '!' && ch <= ';' && ch != '"') return true;  // 0-9 and !#%&'()*+,-./
-    if (ch >= '=' && ch <= 'Z' && ch != '>') return true;  // A-Z and =?@
-    if (ch >= '_' && ch <= 'z' && ch != '`') return true;  // a-z and _
+  return is_numeric(c) || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F');
+}
+
+__device__ bool skip_and_validate_special(string_view::const_iterator& iter,
+                                          string_view::const_iterator end,
```

Comment:
```suggestion
__device__ bool skip_and_validate_special(string_view::const_iterator iter,
                                          string_view::const_iterator end,
```

Reply:
This is actually being modified inside this function as we may skip characters such as hex encoded characters or unicode. I also need to return validity information. I originally had this returning the iterator after it was modified, but switched to this when I also needed validity. Do you think this is easier to understand returning a tuple?

Reply:
Implemented this. Your question tells me that it isn't clear what is happening like if it is a returned value.